File: server/lib/mongodb.ts
import mongoose from "mongoose";

type MongoCache = {
  conn: typeof mongoose | null;
  promise: Promise<typeof mongoose> | null;
};

const globalWithMongoose = globalThis as typeof globalThis & {
  mongoose?: MongoCache;
};

let cached = globalWithMongoose.mongoose;

if (!cached) {
  cached = globalWithMongoose.mongoose = { conn: null, promise: null };
}

function getMongoUri(): string | null {
  const uri = process.env.MONGO_URI?.trim() || process.env.MONGODB_URI?.trim();
  return uri && uri.length > 0 ? uri : null;
}

export const connectDB = async () => {
  const mongoUri = getMongoUri();

  if (!mongoUri) {
    console.warn("[mongodb] MONGO_URI/MONGODB_URI not set — skipping DB connection");
    return null;
  }

  if (cached!.conn) return cached!.conn;

  if (!cached!.promise) {
    cached!.promise = mongoose.connect(mongoUri, {
      // Serverless functions benefit from smaller pools to avoid connection bursts.
      maxPoolSize: 10,
      serverSelectionTimeoutMS: 5000,
      connectTimeoutMS: 10000,
      socketTimeoutMS: 30000,
    });
  }

  cached!.conn = await cached!.promise;
  return cached!.conn;
};

File: PromptWars_Aetherion_project/PromptWars_Aetherion/db/mongodb.py
import os
from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorDatabase

_client: AsyncIOMotorClient | None = None
_db_cache: dict[str, AsyncIOMotorDatabase] = {}


def _get_mongo_uri() -> str | None:
    uri = (os.environ.get("MONGO_URI") or os.environ.get("MONGODB_URI") or "").strip()
    return uri if uri else None


def get_default_db_name() -> str:
    return os.environ.get("MONGO_DB_NAME", "promptwars")


async def get_db(db_name: str | None = None) -> AsyncIOMotorDatabase | None:
    global _client
    uri = _get_mongo_uri()
    if not uri:
        print("[mongodb] MONGO_URI/MONGODB_URI not set — skipping DB connection")
        return None
    resolved_db_name = (db_name or get_default_db_name()).strip() or "promptwars"
    if resolved_db_name in _db_cache:
        return _db_cache[resolved_db_name]
    if _client is None:
        _client = AsyncIOMotorClient(
            uri,
            maxPoolSize=10,
            serverSelectionTimeoutMS=5000,
            connectTimeoutMS=10000,
            socketTimeoutMS=30000,
        )
    db = _client[resolved_db_name]
    _db_cache[resolved_db_name] = db
    return db
